File: problems/687_longest_univalue_path/Solution.java
class Solution {
    int result = 0;
    public int longestUnivaluePath(TreeNode root) {
        dfs(root);
        return result;
    }
    private int dfs(TreeNode node) {
        if (node == null) return 0;
        int left = dfs(node.left);
        int right = dfs(node.right);
        int leftPath = (node.left != null && node.left.val == node.val) ? left + 1 : 0;
        int rightPath = (node.right != null && node.right.val == node.val) ? right + 1 : 0;
        result = Math.max(result, leftPath + rightPath);
        return Math.max(leftPath, rightPath);
    }
}

File: problems/687_longest_univalue_path/solution.py
from shared.python.data_structures import TreeNode


class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        self.result = 0
        def dfs(node):
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            left_path = left + 1 if node.left and node.left.val == node.val else 0
            right_path = right + 1 if node.right and node.right.val == node.val else 0
            self.result = max(self.result, left_path + right_path)
            return max(left_path, right_path)
        dfs(root)
        return self.result
